### ISSUE TITLE
_solib directory placement within runfiles misses external solibs

### ISSUE THREAD

**Issue by jonasschneider** [bazelbuild/bazel#6542]
### Description of the problem / feature request:

I'm trying to build a plug-in `.so` to be later dynamically loaded from runfiles. However, I'm running into an issue with the case where that `.so` uses `linkshared=1` and itself depends on some other libraries (I can reproduce it with both cc_library and cc_import deps). The rpaths set in the .so can get out of sync with the actual filesystem tree, causing load-time link errors. The core issue seems to be the relative path from the .so within the runfiles to the `_solib_{cpu}` directory within the runfiles.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

From an external workspace, build a link_shared=True cc_binary A that links against a cc_library B also in the external project. Then include A in `data` for a target C in the main workspace. Running C shows that A is visible while B is not, leading to a crash when trying to load A. [Repro repo here.](https://github.com/jonasschneider/bazel-solib-bug)

### What operating system are you running Bazel on?

Ubuntu 16.04.5 LTS, x86_64 (but also repros on OS X)

### What's the output of `bazel info release`?

release 0.18.0

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

https://github.com/jonasschneider/bazel-solib-bug

###  Have you found anything relevant by searching the web?

Possible related to https://github.com/bazelbuild/bazel/issues/6481 ?

### Any other information, logs, or outputs that you want to share?

Demonstrating the issue (see [full repro repo](https://github.com/jonasschneider/bazel-solib-bug)):
```
$ cat main.sh
script="${BASH_SOURCE[0]}"
echo "Accessing libplugin.so via .runfiles/__main__/external/child:"
ldd $script.runfiles/__main__/external/child/libplugin.so | grep appcode
echo "Accessing libplugin.so via .runfiles/child:"
ldd $script.runfiles/child/libplugin.so | grep appcode

$ bazel run //:main
[...]
Accessing libplugin.so via .runfiles/__main__/external/child:
	libappcode.so => /root/.cache/bazel/_bazel_root/665bedfccc4ce93904668a7f2f06c5ca/execroot/__main__/bazel-out/k8-fastbuild/bin/main.runfiles/__main__/external/child/../../_solib_k8/_U@child_S_S_Clibplugin.so___Uexternal_Schild/libappcode.so (0x00007f19c3cb8000)
Accessing libplugin.so via .runfiles/child:
	libappcode.so => /root/.cache/bazel/_bazel_root/665bedfccc4ce93904668a7f2f06c5ca/execroot/__main__/bazel-out/k8-fastbuild/bin/main.runfiles/child/../../_solib_k8/_U@child_S_S_Clibplugin.so___Uexternal_Schild/libappcode.so (0x00007f4e4454b000)

$ bazel run //subdir:main
[...]
Accessing libplugin.so via .runfiles/__main__/external/child:
	libappcode.so => /root/.cache/bazel/_bazel_root/665bedfccc4ce93904668a7f2f06c5ca/execroot/__main__/bazel-out/k8-fastbuild/bin/subdir/main.runfiles/__main__/external/child/../../_solib_k8/_U@child_S_S_Clibplugin.so___Uexternal_Schild/libappcode.so (0x00007f382cc13000)
Accessing libplugin.so via .runfiles/child:
	libappcode.so => not found
```

(In my real use-case, libplugin.so is a Python extension module, which in turn depends on a third-party library libappcode.so that's only available as a .so, hence libplugin.so can't be linked fully statically.)

Checking with `patchelf` reveals the source of the loading issue, the rpath of `libplugin.so`:
```
$ patchelf --print-rpath bazel-bin/external/child/libplugin.so
$ORIGIN/../../_solib_k8/_U@child_S_S_Clibplugin.so___Uexternal_Schild
```
It looks like the rpath assumes that libplugin.so is being loaded using the `__main__/external` location. It is correct for that case (a modified test case that also moves `libplugin.so` into a subdirectory within `child` indeed modifies the rpath correctly). However it breaks when both of these conditions are true:
1. `libplugin.so` is loaded using the new `.runfiles/child` syntax instead of `.runfiles/__main__/external/child`
2. The main executable is not at the top of the workspace (since otherwise the rpath "accidentally" resolves to the _solib path _outside_ of the runfiles, which happens to work as well but is brittle)

This seems like a bug -- does that mean there should be two rpath entries, one for each loading location? Or am I doing something wrong here?

**Comment by vsbus** [comment]
Hey @jonasschneider , I am hitting the same issue in my project where generated rpath is relative to `.runfiles/<workspace>/external`. If we have two locations is will solve the problem: `.runfiles/<workspace>/external/` and `.runfiles/`.

Currently, I am adding 2nd rpath hardcoding linkopts like to on of the libraries I import (not I add 2 rpaths b/c I have to import them into 2 different projects):
```
linkopts = [
        "-Wl,-rpath=$$ORIGIN/../../workspace_foo/_solib_local/",
        "-Wl,-rpath=$$ORIGIN/../../__main__/_solib_k8/",
    ],
```
but this doesn't scale as I can't automatically compute `workspace` and `solib` dir names.

I see two solutions for now:
1. automatically add one more path by bazel.
2. expose `solib_dir_path` via special bazel functon or Make Variable that we can use so target project computes additional rpath correctly.

I would be happy to learn if other solutions exist and what could be done to solve this problem. Any ideas?

**Comment by fmeum** [comment]
I would expect this to be fixed in 6.2.0, but haven't tested it.

**Comment by fmeum** [comment]
Definitely fixed in Bazel 7.2.1 (reproducer requires `--noenable_bzlmod`).
